features/atcryptoauth/mbedtls_atca_engine.cpp

```diff
@@ -0,0 +1,355 @@
+{
+    ATCAKey * key = NULL;
+    ATCAError err = ATCA_SUCCESS;
+    ATCADevice * device = ATCAFactory::GetDevice( err );
+    
+    if ( ctx == NULL )
+        return( -1 );
+
+    if ( err != ATCA_SUCCESS )
+    {
+        assert( device == NULL );
+        return( -1 );
+    }
+    key = device->GetKeyToken( keyId, err );
+
+    static const mbedtls_pk_info_t atca_pk_info =
+    {
+        MBEDTLS_PK_OPAQUE,
+        "ATCA",
+        NULL,
+        atca_can_do_func,
+        NULL,
+        atca_verify_func,
+        atca_sign_func,
+        NULL,
+        NULL,
+        NULL,
+        NULL,
+        atca_ctx_free,
+        NULL
+    };
+
+
+    ctx->pk_ctx = (void *)key;
+    ctx->pk_info = &atca_pk_info;
+
+    return( 0 );
+}
+
+/** Setup transparent PK context using Public key from ATCA crypto engine.
+ *
+ * @param ctx       Key context
+ * @param keyId     Private key Id representing a hardware Key.
+ *
+ * @retval          0 if successful, or appropriate Mbed TLS error code.
+ */
+int mbedtls_atca_transparent_pk_setup( mbedtls_pk_context * ctx, ATCAKeyID keyId )
+{
+    ATCAKey * key = NULL;
+    ATCAError err = ATCA_SUCCESS;
+    ATCADevice * device = ATCAFactory::GetDevice( err );
+    if ( err != ATCA_SUCCESS )
+    {
+        assert( device == NULL );
+        return MBEDTLS_ERR_PK_HW_ACCEL_FAILED;
+    }
+
+    /* Import public key from binary */
+    static mbedtls_ecp_keypair ecp_key;
+    mbedtls_ecp_keypair_init(&ecp_key);
+    int ret = mbedtls_ecp_group_load(&ecp_key.grp, MBEDTLS_ECP_DP_SECP256R1);
+    if ( ret != 0 )
+        return( ret );
+    key = device->GetKeyToken( keyId, err );
+    if ( ret != 0 )
+    {
+        printf(" failed\n  !  to retrieve Public key from ATCA508A\n\n" );
+        return( ret );
+    }
+    uint8_t pubkey_asn1[ATCA_ECC_ECC_PK_LEN + 10];
+    size_t asn_len = 0;
+    ret = ecc_key_to_asn1( key->GetPubKey(), pubkey_asn1, sizeof(pubkey_asn1), &asn_len);
+    delete key;
+    if ( ret != 0 )
+    {
+        printf( " failed\n  !  to ASN.1 encode ECDSA Public key binary\n\n" );
+        return( ret );
+    }
+    ret = mbedtls_ecp_point_read_binary(&ecp_key.grp, &ecp_key.Q, pubkey_asn1, asn_len );
+    if ( ret != 0 )
+    {
+        printf( " failed\n  !  Failed to read ecp key from binary\n\n" );
+        return( ret );
+    }
+    ctx->pk_info = mbedtls_pk_info_from_type( MBEDTLS_PK_ECKEY );
+    ctx->pk_ctx = &ecp_key;
+    return( 0 );
+}
+
+/** Encode ECC Public key in ASN.1 format.
+ *
+ *  @param ecc_pk       ECC Public key input.
+ *  @param asn_out      ASN.1 out buffer.
+ *  @param asn_out_len  Out buffer length.
+ *  @return             0 on success, -1 on failure
+ */
+int ecc_key_to_asn1( uint8_t * ecc_pk, uint8_t * asn_out, size_t asn_len, size_t * asn_out_len )
+{
+    int ret = 0;
+    mbedtls_ecp_keypair ecp_key;
+    mbedtls_ecp_keypair_init(&ecp_key);
+
+    if (mbedtls_ecp_group_load(&ecp_key.grp, MBEDTLS_ECP_DP_SECP256R1) != 0)
+    {
+        goto cleanup;
```

Comment:
This branch will return 0, signalling a success.
Please initialize ret to -1 in the beginning, and add an assignment here.